File: win-prog/ProgWinVB/Tapping Into the Keyboard/SysInfoKeyboard/SysInfoKeyboard.vb
'------------------------------------------------
' SysInfoKeyboard.vb (c) 2002 by Charles Petzold
'------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class SysInfoKeyboard
    Inherits SysInfoReflection

    Shared Shadows Sub Main()
        Application.Run(new SysInfoKeyboard())
    End Sub

    Sub New()
        Text = "System Information: Keyboard"
    End Sub

    Protected Overrides Sub OnKeyDown(ByVal kea As KeyEventArgs)
        Dim pt As Point = AutoScrollPosition
        pt.X = -pt.X
        pt.Y = -pt.Y

        Select kea.KeyCode
            Case Keys.Right
                If (kea.Modifiers And Keys.Control) = Keys.Control Then
                    pt.X += ClientSize.Width
                Else
                    pt.X += Font.Height
                End If

            Case Keys.Left
                If (kea.Modifiers And Keys.Control) = Keys.Control Then
                    pt.X -= ClientSize.Width
                Else
                    pt.X -= Font.Height
                End If

            Case Keys.Down
                pt.Y += Font.Height

            Case Keys.Up
                pt.Y -= Font.Height

            Case Keys.PageDown
                pt.Y += Font.Height * (ClientSize.Height \ Font.Height)

            Case Keys.PageUp
                pt.Y -= Font.Height * (ClientSize.Height \ Font.Height)

            Case Keys.Home
                pt = Point.Empty

            Case Keys.End
                pt.Y = 1000000
        End Select

        AutoScrollPosition = pt
    End Sub
End Class

File: win-prog/ProgWinCS/Tapping Into the Keyboard/SysInfoKeyboard/SysInfoKeyboard.cs
//----------------------------------------------
// SysInfoKeyboard.cs © 2001 by Charles Petzold
//----------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class SysInfoKeyboard: SysInfoReflection
{
     public new static void Main()
     {
          Application.Run(new SysInfoKeyboard());
     }
     public SysInfoKeyboard()
     {
          Text = "System Information: Keyboard";
     }
     protected override void OnKeyDown(KeyEventArgs kea)
     {
          Point pt = AutoScrollPosition;

          pt.X = -pt.X;
          pt.Y = -pt.Y;

          switch(kea.KeyCode)
          {
          case Keys.Right:
               if ((kea.Modifiers & Keys.Control) == Keys.Control)
                    pt.X += ClientSize.Width;
               else
                    pt.X += Font.Height;
               break;

          case Keys.Left:      
               if ((kea.Modifiers & Keys.Control) == Keys.Control)
                    pt.X -= ClientSize.Width;
               else
                    pt.X -= Font.Height;
               break;
               
          case Keys.Down:      pt.Y += Font.Height;        break;
          case Keys.Up:        pt.Y -= Font.Height;        break;
          case Keys.PageDown:  
               pt.Y += Font.Height * (ClientSize.Height / Font.Height);  
               break;
          case Keys.PageUp:    
			   pt.Y -= Font.Height * (ClientSize.Height / Font.Height);
			   break;
          case Keys.Home:      pt    = Point.Empty;        break;
          case Keys.End:       pt.Y  = 1000000;            break;
          }
          AutoScrollPosition = pt;
     }
}
